fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -980,69 +958,130 @@ ACTOR Future<bool> checkExclusion(Database db,
-	state bool excludedAddressesContainsStorageRole = false;
+	state std::map<std::string, std::vector<std::string>> parsedLocalities;
+	// Convert the passed localities into a map of vectors to make it easier to check if a process
+	// is excluded by locality.
+	for (const auto& locality : localities) {
+		std::pair<std::string, std::string> locality_key_value = decodeLocality(locality);
+		if (locality_key_value.first == "") {
+			continue;
+		}
+
+		parsedLocalities[locality_key_value.first].push_back(locality_key_value.second);
+	}
+
+	// If the exclusion marks the excluded processes as failed and the provided addresses are not empty,
+	// we will copy the input addresses into a new vector to not modify the input vector when additional
+	// addresses are added. We do this to get the excluded addresses when locality based exclusions are used
+	// with the failed flag, in this case the checkSafeExclusions will be called with the excluded addresses
+	// to verify if the exclusion can be done. In case of locality based exclusions the addresses should
+	// be empty and if the addresses are not empty, we are not using the locality based exclusion. So this
+	// step is a safeguard for future changes where the provided addresses are not empty and locality based
+	// exclusions are used.
+	state std::vector<AddressExclusion> excludedAddresses;
+	if (markFailed && !addresses.empty()) {
+		excludedAddresses = addresses;
+	}
 
 	try {
 		for (auto proc : processesMap.obj()) {
 			StatusObjectReader process(proc.second);
-			std::string addrStr;
-			if (!process.get("address", addrStr)) {
-				*msg = ManagementAPIError::toJsonString(false, markFailed ? "exclude failed" : "exclude", errorString);
-				return false;
+			StatusObjectReader localityObj;
+			bool hasLocalities = process.get("locality", localityObj);
+
+			bool excluded = process.has("excluded") && process.last().get_bool();
+			// If the process is already excluded based on the process status, we don't have to check if the process is
+			// also excluded by address or locality.
+			if (!excluded) {
+				// First check if the process is excluded based on the exclusions.
+				std::string addrStr;
+				if (!process.get("address", addrStr)) {
+					*msg =
+					    ManagementAPIError::toJsonString(false, markFailed ? "exclude failed" : "exclude", errorString);
+					return false;
+				}
+				NetworkAddress addr = NetworkAddress::parse(addrStr);
+				excluded = addressExcluded(exclusions, addr);
+
+				// If the process is not already excluded and the parsed localites have at least one entry,
+				// check if the process is excluded by localities.
+				if (!excluded && hasLocalities && !parsedLocalities.empty()) {
+					// Iterate over all excluded localites and check if the process is excluded based on the locality.
+					for (std::map<std::string, std::vector<std::string>>::iterator it = parsedLocalities.begin();
+					     it != parsedLocalities.end();
+					     ++it) {
```

Comment:
```suggestion
					for (const auto& parsedLocality : parsedLocalities) {
```

Some benefits:
- more readable
- no copy

You will still have to do `parsedLocality.first` and `parsedLocality.second` instead of `it->first` and `it->second`. If you want to make it even more readable, then you can try:

```suggestion
					for (const auto& [localityKey, localityVal] : parsedLocalities) {
```

Reply:
Oops. I merged without checking this comment.

Reply:
No worries, it was a minor so I approved and left a comment. I'll message @johscheuer in case he wants to make a follow-up PR.